src/internal/task/etcd_task_service.go

```diff
@@ -0,0 +1,340 @@
+			if err := ed.taskCol.ReadWrite(stm).Delete(taskKey); err != nil {
+				return err
+			}
+			return ed.claimCol.ReadWrite(stm).Delete(taskKey)
+		}); err != nil {
+			fmt.Printf("errored deleting task %v: %v\n", taskKey, err)
+		}
+	}()
+	var result *types.Any
+	if err := ed.taskCol.ReadOnly(ctx).WatchOneF(taskKey, func(e *watch.Event) error {
+		if e.Type == watch.EventDelete {
+			return errors.New("task was deleted while waiting for results")
+		}
+		var key string
+		task := &Task{}
+		if err := e.Unmarshal(&key, task); err != nil {
+			return err
+		}
+		if task.State == State_RUNNING {
+			return nil
+		}
+		if task.State == State_FAILURE {
+			return errors.New(task.Reason)
+		}
+		result = task.Output
+		return errutil.ErrBreak
+	}); err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+type CollectFunc func(int, *types.Any, error) error
+
+func (ed *etcdDoer) DoBatch(ctx context.Context, anys []*types.Any, cb CollectFunc) error {
+	if len(anys) == 0 {
+		return nil
+	}
+	batchPrefix := path.Join(ed.groupID, uuid.NewWithoutDashes())
+	idxMap := make(map[string]int)
+	for i, any := range anys {
+		// TODO: include any type in hash?
+		sum := pachhash.Sum(any.Value)
+		taskID := pachhash.EncodeHash(sum[:])
+		taskKey := path.Join(batchPrefix, taskID)
+		task := &Task{ID: taskID, Input: any, State: State_RUNNING}
+		if err := ed.renewer.Put(ctx, taskKey, task); err != nil {
+			return err
+		}
+		idxMap[taskKey] = i
+	}
+	defer func() {
+		if _, err := col.NewSTM(ctx, ed.etcdClient, func(stm col.STM) error {
+			if err := ed.taskCol.ReadWrite(stm).DeleteAllPrefix(batchPrefix); err != nil {
+				return err
+			}
+			return ed.claimCol.ReadWrite(stm).DeleteAllPrefix(batchPrefix)
+		}); err != nil {
+			fmt.Printf("errored deleting all tasks in a batch: %v\n", err)
+		}
+	}()
+	return ed.taskCol.ReadOnly(ctx).WatchOneF(batchPrefix, func(e *watch.Event) error {
+		if e.Type == watch.EventDelete {
+			return errors.New("task was deleted while waiting for results")
+		}
+		var key string
+		task := &Task{}
+		if err := e.Unmarshal(&key, task); err != nil {
+			return err
+		}
+		if task.State == State_RUNNING {
+			return nil
+		}
+		var err error
+		if task.State == State_FAILURE {
+			err = errors.New(task.Reason)
+		}
+		if err := cb(idxMap[key], task.Output, err); err != nil {
+			return err
+		}
+		delete(idxMap, key)
+		if len(idxMap) == 0 {
+			return errutil.ErrBreak
+		}
+		return nil
+	})
+}
+
+func (es *etcdService) Maker(namespace string) Maker {
+	namespaceEtcd := newNamespaceEtcd(es.etcdClient, es.etcdPrefix, namespace)
+	return newEtcdMaker(namespaceEtcd)
+}
+
+type Maker interface {
+	Make(ctx context.Context, cb ProcessFunc) error
+}
+
+type etcdMaker struct {
+	*namespaceEtcd
+}
+
+func newEtcdMaker(namespaceEtcd *namespaceEtcd) Maker {
+	return &etcdMaker{
+		namespaceEtcd: namespaceEtcd,
+	}
+}
+
+type ProcessFunc func(context.Context, *types.Any) (*types.Any, error)
```

Comment:
This should be a type alias, it improves the compiler errors when you pass the wrong type.